patroni/utils.py

```diff
@@ -448,3 +449,40 @@ def data_directory_is_empty(data_dir):
     if not os.path.exists(data_dir):
         return True
     return all(os.name != 'nt' and (n.startswith('.') or n == 'lost+found') for n in os.listdir(data_dir))
+
+
+def is_cluster_healthy(patroni, cluster):
+    if cluster.leader:
+        leader_name = cluster.leader.member.name
+    else:
+        logger.warning('cluster is not healthy: no leader')
+        return False
+    for m in cluster.members:
+        if m.name == leader_name:
+            leader_tl = m.data.get('timeline', '')
+            # sanity check
+            if m.data.get('role', '') != 'master' or m.data.get('state', '') != 'running':
+                logger.warning('cluster is not healthy: leader does not have role master or is not in state running')
+                return False
+            # check replication data from leader and make sure there are as many rows as replicas
+            try:
+                response = patroni.request(m, timeout=2, retries=0)
+                data = json.loads(response.data.decode('utf-8'))
+            except Exception as e:
+                logger.warning("Request failed to %s: GET %s (%s)", m.name, m.api_url, e)
+            if not 'replication' in data or len(data['replication']) + 1 != len(cluster.members):
+                logger.warning('cluster is not healthy: not all members take part in replication')
+                return False
+    if cluster.config:
+        maximum_lag_on_failover = cluster.config.data.get('maximum_lag_on_failover', 0)
+    for m in cluster.members:
+        if m.name == leader_name:
+            continue
+        if m.data.get('timeline', '') != leader_tl or int(m.data.get('lag', 0)) > maximum_lag_on_failover:
+            logger.warning('cluster is not healthy: timeline mismatch in member %s', m.name)
```

Comment:
The warning is misleading, while in fact it could be on the same timeline but replication is lagging.
It is also possible that the replica is tagged with `nofailover` and  `recovery_min_apply_delay` is set to some high values.

Reply:
About the former, I would say the cluster isn't very healthy in this case, but the warning message could be improved to add the max log thing.

About the latter, I will need to think a bit more about this; maybe in this case the user cannot expect this feature to work correctly

Reply:
The second paragraph is addressed by #2089/#2108